// ngm-cli/tasks/bundle-umd.task.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const utils_1 = require("../utils");
const webpack_umd_config_1 = require("../models/webpack-umd.config");
const webpack = require('webpack');
// todo: move to constants and make it configurable
const bundlesDir = 'bundles';
// Configure build and output;
let lastHash = null;
const webpackOutputOptions = {
    colors: true,
    chunks: true,
    modules: false,
    reasons: false,
    chunkModules: false
};
// export function bundleUmd(dir, moduleConf, minify) {
function bundleUmd({ src, dist, name, main, tsconfig, minify }) {
    const config = webpack_umd_config_1.getWebpackConfig({
        name: !minify ? `${name}.umd` : `${name}.umd.min`,
        root: path.resolve(utils_1.ROOT, src),
        entry: path.resolve(utils_1.ROOT, src, main),
        output: path.resolve(dist, bundlesDir),
        tsconfig: tsconfig
    });
    if (minify) {
        config.optimization.minimize = true;
    }
    const webpackCompiler = webpack(config);
    return new Promise((resolve, reject) => {
        webpackCompiler.run((err, stats) => {
            if (err) {
                if (stats) {
                    process.stdout.write(stats.toString(webpackOutputOptions) + '\n');
                }
                return reject(err);
            }
            if (stats.hasErrors()) {
                process.stdout.write(stats.toString(webpackOutputOptions) + '\n');
            }
            return stats.hasErrors() ? reject() : resolve();
        });
    });
}
exports.bundleUmd = bundleUmd;
