// ngm-cli/tasks/bundle-es2015.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = require('path');
const execa = require('execa');
const rollup = require('rollup');
const resolve = require('rollup-plugin-node-resolve');
const commonjs = require('rollup-plugin-commonjs');
const inline_resources_1 = require("../helpers/inline-resources");
const rollup_globals_1 = require("../models/rollup.globals");
const bundlesDir = 'bundles';
function bundleEs2015(config) {
    return __awaiter(this, void 0, void 0, function* () {
        yield inline_resources_1.inlineResources(config.tmp);
        yield execa('ngc', ['-p', config.tmp], { preferLocal: true });
        return rollup.rollup({
            input: path.resolve(config.tmp, 'dist-es2015', config.input.replace('.ts', '')),
            external: Object.keys(rollup_globals_1.default),
            plugins: [
                resolve({
                    module: true,
                    main: true
                }),
                commonjs({
                    include: 'node_modules/**',
                })
            ],
            onwarn: warning => {
                const skip_codes = [
                    'THIS_IS_UNDEFINED',
                    'MISSING_GLOBAL_NAME'
                ];
                if (skip_codes.indexOf(warning.code) != -1)
                    return;
                console.error(warning);
            }
        }).then(bundle => bundle.write({
            file: path.resolve(config.dist, bundlesDir, config.name + '.es2015.js'),
            name: config.name,
            format: 'es',
            sourcemap: true
        }));
    });
}
exports.bundleEs2015 = bundleEs2015;
